pkg/apis/crd/v1alpha2/types.go

```diff
@@ -280,3 +270,63 @@ type ExternalIPPoolList struct {
 
 	Items []ExternalIPPool `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPPool defines one or multiple IP sets that can be used for flexible IPAM feature. For instance, the IPs can be
+// allocated to pods according to IP pool specified in deployment annotation.
+type IPPool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the ExternalIPPool.
+	Spec IPPoolSpec `json:"spec"`
+
+	// Most recently observed status of the pool.
+	Status IPPoolStatus `json:"status"`
+}
+
+type IPPoolSpec struct {
+	// IP Version for this IP pool - either 4 or 6
+	IPVersion int `json:"ipVersion"`
+	// The IP ranges of this IP pool, e.g. 10.10.0.0/24, 10.10.10.2-10.10.10.20, 4001::12-4001::30.
+	IPRanges []IPRange `json:"ipRanges"`
+	// Vlan ID for this IP pool. Default is 0. String-typed for sake of potential autoselect option.
+	Vlan string `json:"vlan,omitempty"`
```

Comment:
VLAN

Reply:
Do you think VLAN should be an attribute of IPRange, so one IPPool can include multiple subnets?

Reply:
This would increase feature complexity - lets discuss?

Reply:
A potential compromise could be to allow multiple pool annotation per resource (of single IP version). This way we can keep single vlan (and potentially GW) per pool.

Reply:
But why multiple pools per resource is simpler than multiple VLANs per pool?

Reply:
This way we could start with simple feature and increase complexity as need arises

Reply:
Besides VLAN, we also need gateway and prefix length for each subnet, why you like to put them?

I thought about another way of incremental implementation - adding VLAN to IPRange later and then let IPRange VLAN (if set) override IPPool VLAN, but if we need to do that for all VLAN, Gateway, prefix length, that sounds worse (to duplicate all these in both IPPool and IPRange and let IPRange override IPPool).

For user easy consumption, I still think we should (finally) support multiple subnets per IPPool.

Reply:
In my view, an IPPool is a collection of Subnets, and within those Subnets, I can have ranges that e.g. Deployments allocate IPs from. And each Subnet has (can have) a Gateway and a traffic tag (VLAN). We could also have a top-level Traffic Tag setting that applies to all Subnets in the Pool, but we can't have a top-level Gateway. If we have a traffic tag for all IPRanges, I would still want to have the ability to overwrite it per IPRange.
Now I think we already went down the path of having the IPRange Object being either a subnet or range, and we don't have a Subnet object. In this case, we also need a prefix setting per IPRange as don't have a Subnet object below the IPRange object from which one can derive the prefix.

Or as a second option, we could decide that an IP Pool can onky have a single "Top Level" subnet. In this case, we need to be able to allocate multiple IPPools to e.g. deployments so that IPs can be added to deployments as a day 2 operation. 

I would prefer the first option, and I think this matches Jianjun's thoughts.

Reply:
Thanks, I've modified the CRD accordingly

Reply:
I'm OK for VLAN as an attribute of IPRange.